qiskit/quantum_info/synthesis/xx_decompose/decomposer.py

```diff
@@ -0,0 +1,278 @@
+)
+
+from .circuits import apply_reflection, apply_shift, canonical_xx_circuit
+from .utilities import EPSILON
+from .polytopes import XXPolytope
+
+
+def _average_infidelity(p, q):
+    """
+    Computes the infidelity distance between two points p, q expressed in positive canonical
+    coordinates.
+    """
+
+    a0, b0, c0 = p
+    a1, b1, c1 = q
+
+    return 1 - 1 / 20 * (
+        4
+        + 16
+        * (
+            math.cos(a0 - a1) ** 2 * math.cos(b0 - b1) ** 2 * math.cos(c0 - c1) ** 2
+            + math.sin(a0 - a1) ** 2 * math.sin(b0 - b1) ** 2 * math.sin(c0 - c1) ** 2
+        )
+    )
+
+
+class XXDecomposer:
+    """
+    A class for optimal decomposition of 2-qubit unitaries into 2-qubit basis gates of XX type
+    (i.e., each locally equivalent to CAN(alpha, 0, 0) for a possibly varying alpha).
+
+    Args:
+        euler_basis: Basis string provided to OneQubitEulerDecomposer for 1Q synthesis.
+            Defaults to "U".
+        embodiments: A dictionary mapping interaction strengths alpha to native circuits which
+            embody the gate CAN(alpha, 0, 0). Strengths are taken so that pi/2 represents the class
+            of a full CX.
+        backup_optimizer: If supplied, defers synthesis to this callable when XXDecomposer
+            has no efficient decomposition of its own. Useful for special cases involving 2 or 3
+            applications of XX(pi/2), in which case standard synthesis methods provide lower
+            1Q gate count.
+
+    NOTE: If embodiments is not passed, or if an entry is missing, it will be populated as needed
+        using the method _default_embodiment.
+    """
+
+    def __init__(
+        self,
+        euler_basis: str = "U",
+        embodiments: Optional[dict] = None,
+        backup_optimizer: Optional[Callable] = None,
+    ):
+        self._decomposer1q = Optimize1qGatesDecomposition(ONE_QUBIT_EULER_BASIS_GATES[euler_basis])
+        self.gate = RZXGate(np.pi / 2)
+        self.embodiments = embodiments if embodiments is not None else {}
+        self.backup_optimizer = backup_optimizer
+
+        self._check_embodiments()
+
+    @staticmethod
+    def _default_embodiment(strength):
+        """
+        If the user does not provide a custom implementation of XX(strength), then this routine
+        defines a default implementation using RZX.
+        """
+        xx_circuit = QuantumCircuit(2)
+
+        # NOTE: One could branch here on `strength == np.pi / 2` and decide to use a CX-based
+        #       circuit in this one case where it's available.
+        xx_circuit.h(0)
+        xx_circuit.rzx(strength, 0, 1)
+        xx_circuit.h(0)
+
+        return xx_circuit
+
+    def _check_embodiments(self):
+        """
+        Checks that `self.embodiments` is populated with legal circuit embodiments: the key-value
+        pair (angle, circuit) satisfies Operator(circuit) approx RXX(angle).to_matrix().
+        """
+
+        for angle, embodiment in self.embodiments:
```

Comment:
```suggestion
        for angle, embodiment in self.embodiments.items():
```

probably good to add a test too with a wrong embodiment to show it catches the error